src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1687,42 +1687,48 @@
          candidates (->> (partitioning/get-partition-names table-prefix)
                          (filter expired?)
                          sort)
+         ts-string (.toString (now) "yyyyMMddHHssSSS")
          drop-one (fn [table]
                     (update-lock-status status-key inc)
                     (try!
                       (if just-detach?
-                        (jdbc/do-commands-outside-txn
-                          (format "alter table %s detach partition %s concurrently" table-prefix table))
-                        (jdbc/do-commands
-                          (format "drop table if exists %s cascade" table)))
+                        (let [detached-name (format "%s_detached_%s" table ts-string)]
+                          (jdbc/do-commands-outside-txn
+                            (format "alter table if exists %s detach partition %s concurrently" table-prefix table)
+                            (format "alter table if exists %s rename to %s" table detached-name))
```

Comment:
This is an attempt to reduce the window in which a new store thread fails because a new partition can't be created due to the existing table with same name.

It's messy, and I don't know how safe it is, in terms of time, what happens to indices, serial...